G++ doesn't value-initialize all members of non-trivial type in default argument

 http://gcc.gnu.org/bugzilla/show_bug.cgi?id=50793

--- gcc-4_4-branch/gcc/cp/ChangeLog	2011/10/19 22:16:17	180222
+++ gcc-4_4-branch/gcc/cp/ChangeLog	2011/10/19 22:21:15	180223
@@ -1,3 +1,8 @@
+2011-10-19  Jason Merrill  <jason@redhat.com>
+
+	PR c++/50793
+	* tree.c (bot_manip): Propagate AGGR_INIT_ZERO_FIRST.
+
 2011-10-13  Jason Merrill  <jason@redhat.com>
 
 	PR c++/50618
--- gcc-4_4-branch/gcc/cp/tree.c	2011/10/19 22:16:17	180222
+++ gcc-4_4-branch/gcc/cp/tree.c	2011/10/19 22:21:15	180223
@@ -1536,7 +1536,11 @@
       tree u;
 
       if (TREE_CODE (TREE_OPERAND (t, 1)) == AGGR_INIT_EXPR)
-	u = build_cplus_new (TREE_TYPE (t), TREE_OPERAND (t, 1));
+	{
+	  u = build_cplus_new (TREE_TYPE (t), TREE_OPERAND (t, 1));
+	  if (AGGR_INIT_ZERO_FIRST (TREE_OPERAND (t, 1)))
+	    AGGR_INIT_ZERO_FIRST (TREE_OPERAND (u, 1)) = true;
+	}
       else
 	u = build_target_expr_with_type (TREE_OPERAND (t, 1), TREE_TYPE (t));
 
--- gcc-4_4-branch/gcc/testsuite/ChangeLog	2011/10/19 22:16:17	180222
+++ gcc-4_4-branch/gcc/testsuite/ChangeLog	2011/10/19 22:21:15	180223
@@ -1,3 +1,8 @@
+2011-10-19  Jason Merrill  <jason@redhat.com>
+
+	PR c++/50793
+	* g++.dg/init/value9.C: New.
+
 2011-10-13  Jason Merrill  <jason@redhat.com>
 
 	PR c++/50618
--- /dev/null
+++ gcc-4_4-branch/gcc/testsuite/g++.dg/init/value9.C	2011/10/19 22:21:15	180223
@@ -0,0 +1,32 @@
+// PR c++/50793
+// { dg-do run }
+
+struct NonTrivial
+{
+  NonTrivial() { }
+};
+
+struct S
+{
+  NonTrivial nt;
+  int i;
+};
+
+int f(S s)
+{
+  s.i = 0xdeadbeef;
+  return s.i;
+}
+
+int g(S s = S())
+{
+  return s.i;
+}
+
+int main()
+{
+  f(S());  // make stack dirty
+
+  if ( g() )
+    __builtin_abort();
+}
